spharpy/classes/ambisonics.py

```diff
@@ -0,0 +1,220 @@
+            ``'real'``. The default is ``'real'``.
+        normalization : str
+            Normalization convention, either ``'n3d'``, ``'maxN'`` or
+            ``'sn3d'``. The default is ``'n3d'``.
+            (maxN is only supported up to 3rd order)
+        channel_convention : str
+            Channel ordering convention, either ``'acn'`` or ``'fuma'``.
+            The default is ``'acn'``.
+            (FuMa is only supported up to 3rd order)
+        condon_shortley : bool or str, optional
+            Whether to include the Condon-Shortley phase term. If ``True`` or
+            ``'auto'``, Condon-Shortley is included, if ``False`` it is not
+            included. The default is ``'auto'``.
+        n_samples : int, optional
+            Number of samples of the time signal. Required if domain is
+            ``'freq'``. The default is ``None``, which assumes an even number
+            of samples if the data is provided in the frequency domain.
+        domain : ``'time'``, ``'freq'``, optional
+            Domain of data. The default is ``'time'``
+        fft_norm : str, optional
+            The normalization of the Discrete Fourier Transform (DFT). Can be
+            ``'none'``, ``'unitary'``, ``'amplitude'``, ``'rms'``, ``'power'``,
+            or ``'psd'``. See :py:func:`~pyfar.dsp.fft.normalization` and [#]_
+            for more information. The default is ``'none'``, which is typically
+            used for energy signals, such as impulse responses.
+        comment : str
+            A comment related to `data`. The default is ``None``.
+
+        References
+        ----------
+        .. [#] E.G. Williams, "Fourier Acoustics", (1999), Academic Press
+        .. [#] B. Rafely, "Fundamentals of Spherical Array Processing", (2015),
+               Springer-Verlag
+        .. [#] F. Zotter, M. Frank, "Ambisonics A Practical 3D Audio Theory
+               for Recording, Studio Production, Sound Reinforcement, and
+               Virtual Reality", (2019), Springer-Verlag
+
+        """
+    def __init__(
+            self,
+            data,
+            sampling_rate,
+            n_max,
+            basis_type,
+            normalization,
+            channel_convention,
+            condon_shortley='auto',
+            n_samples=None,
+            domain='time',
+            fft_norm='none',
+            comment="",
+            is_complex=False):
+        """
+        Create SphericalHarmonicSignal with data, and sampling rate.
+        """
+        # set n_max
+        if (n_max < 0) or (n_max % 1 != 0):
+            raise ValueError("n_max must be a positive integer")
+        if not data.shape[-2] >= (n_max + 1) ** 2:
+            raise ValueError('Data has to few sh coefficients for '
+                             f'{n_max=}. Highest possible n_max is '
+                             f'{int(np.sqrt(data.shape[-2]))-1}')
+        self._n_max = int(n_max)
```

Comment:
see comment above. I would suggest

```suggestion
        self._n_max = int(np.round(np.sqrt(data.shape[-2]) - 1))
```

and add a check, if the number of coefficients makes sense, maybe by checking `(self._n_max + 1)**2 == data.shape[-2]`. This would require first calling `Signal.__init__` to make sure data is valid (numeric) and is already a numpy array.